### PR TITLE
New Resource: `azurerm_hybrid_compute_machine_extension`

### PR DESCRIPTION
**Pull request by liuwuliuyun**
New Resource: `azurerm_hybrid_compute_machine_extension`
Swagger: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2022-11-10/HybridCompute.json

Testing evidence will be provided after TC run completes.

### CODE REVIEW

**Comment by liuwuliuyun** [comment]
Try to rerun accTest with para = 1

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks for this @liuwuliuyun - I've left a couple of minor comments here for you to consider.

Also, given that this is known as "Azure Arc" in the portal, we think it might make more sense to rename this along those same lines - `azurerm_arc_machine_extension` perhaps? This would also involve renaming the corresponding  [datasource](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/hybridcompute/hybrid_compute_machine_data_source.go) too. Thanks!

**Comment by liuwuliuyun** [comment]
Agreed on the renaming. Will do renaming and improvements on the datasource first.

File: internal/services/hybridcompute/hybrid_compute_machine_extension_resource.go [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+		"force_update_tag": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"location": commonschema.Location(),
+
+		"protected_settings": {
```
[the remark this answers is not in the document]
Comment (liuwuliuyun) [reply internal/services/hybridcompute/hybrid_compute_machine_extension_resource.go]: Agreed and will change

File: internal/services/hybridcompute/hybrid_compute_machine_extension_resource.go [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: machines.ValidateMachineID,
+		},
+
+		"automatic_upgrade_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Computed: true,
```
[the remark this answers is not in the document]
Comment (liuwuliuyun) [reply internal/services/hybridcompute/hybrid_compute_machine_extension_resource.go]: Try to set a default value to true to see if it passes test without computed tag

File: internal/services/hybridcompute/arc_machine_data_source.go [hunk 3]
```diff
@@ -116,23 +116,23 @@ type ErrorAdditionalInfoModel struct {
...
-type HybridComputeMachineDataSource struct{}
+type ArcMachineDataSource struct{}
 
-var _ sdk.DataSource = HybridComputeMachineDataSource{}
+var _ sdk.DataSource = ArcMachineDataSource{}
 
-func (r HybridComputeMachineDataSource) ResourceType() string {
-	return "azurerm_hybrid_compute_machine"
+func (r ArcMachineDataSource) ResourceType() string {
+	return "azurerm_arc_machine"
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_data_source.go:124]: This is a breaking change, we'd need to add a new data source with the correct name and deprecate this one.

File: internal/services/hybridcompute/arc_machine_extension_resource.go [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+
+		"location": commonschema.Location(),
+
+		"protected_settings": {
+			Type:             pluginsdk.TypeString,
+			Optional:         true,
+			Sensitive:        true,
+			ValidateFunc:     validation.StringIsJSON,
+			DiffSuppressFunc: pluginsdk.SuppressJsonDiff,
+		},
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_extension_resource.go:106]: What can be set in here?

File: internal/services/hybridcompute/arc_machine_extension_resource.go [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"settings": {
+			Type:             pluginsdk.TypeString,
+			Optional:         true,
+			ValidateFunc:     validation.StringIsJSON,
+			DiffSuppressFunc: pluginsdk.SuppressJsonDiff,
+		},
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_extension_resource.go:120]: Same here, what can settings can be set in here?

File: internal/services/hybridcompute/arc_machine_extension_resource.go [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+
+		"type": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_extension_resource.go:127]: We don't expose `type` fields, what values can this have? We should potentially split this out into separate resources.

File: internal/services/hybridcompute/arc_machine_extension_resource.go [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+
+								"message": {
+									Type:     pluginsdk.TypeString,
+									Computed: true,
+								},
+
+								"time": {
+									Type:     pluginsdk.TypeString,
+									Computed: true,
+								},
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_extension_resource.go:191]: Is it possible to have multiple statuses? These properties can probably be flattened and moved up a level

File: internal/services/hybridcompute/arc_machine_extension_resource.go [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+							},
+						},
+					},
+
+					"type": {
+						Type:     pluginsdk.TypeString,
+						Computed: true,
+					},
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_extension_resource.go:199]: Same comment regarding the field `type` we do not usually expose these fields.

File: internal/services/hybridcompute/arc_machine_extension_resource.go [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+					return metadata.MarkAsGone(id)
+				}
+
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			model := resp.Model
+			if model == nil {
+				return fmt.Errorf("retrieving %s: model was nil", id)
+			}
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_extension_resource.go:408]: We don't need to return an error if Model is nil (same applies to Properties) and for readability we assign it to a variable

```suggestion
			if model := resp.Model; model != nil {
				...
				if props := model.Properties; props != nil {
				...
			
```

File: internal/services/hybridcompute/arc_machine_extension_resource_test.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+  name                               = "acctest-hcme-%d"
+  arc_machine_id                     = data.azurerm_arc_machine.test.id
+  location                           = "%s"
+  automatic_upgrade_enabled          = false
+  publisher                          = "Microsoft.Azure.Monitor"
+  type                               = "AzureMonitorLinuxAgent"
+  type_handler_version               = "1.24"
+}
+`, template, data.RandomInteger, data.Locations.Primary)
+}
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_extension_resource_test.go:159]: It looks like this config is missing a fair few properties for it to be complete?

File: internal/services/hybridcompute/arc_machine_extension_resource_test.go [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+func (r HybridComputeMachineExtensionResource) template(data acceptance.TestData) string {
+	d := HybridComputeMachineDataSource{}
+	clientSecret := os.Getenv("ARM_CLIENT_SECRET")
+	randomUUID, _ := uuid.GenerateUUID()
+	password := generateRandomPassword(10)
+	return d.basic(data, clientSecret, randomUUID, password)
+}
```
Comment (stephybun) [inline internal/services/hybridcompute/arc_machine_extension_resource_test.go:183]: For consistency the template should exist within the resource's tests and be referenced by the data source tests, not the other way around.

**Review by stephybun — CHANGES_REQUESTED** [review]

File: internal/services/hybridcompute/arc_machine_data_source.go (hunk 3 above)
Comment (liuwuliuyun) [reply internal/services/hybridcompute/arc_machine_data_source.go:124]: Agreed, I will add this to a sperate data source.

File: internal/services/hybridcompute/arc_machine_extension_resource_test.go (hunk 11 above)
Comment (liuwuliuyun) [reply internal/services/hybridcompute/arc_machine_extension_resource_test.go:183]: Ok, Let me change this.

File: internal/services/hybridcompute/arc_machine_extension_resource_test.go (hunk 10 above)
Comment (liuwuliuyun) [reply internal/services/hybridcompute/arc_machine_extension_resource_test.go:159]: Let me add settings and protected settings example to complete config.

**Comment by liuwuliuyun** [comment]
@stephybun @catriona-m what do you guys think of the naming of resources in the resource provider? Currently we have two solutions. 
1. keeping every resource starts with `azurerm_hybrid_compute_machine*`. This solution has no breaking change for users and it aligns with the RP name.
2. change to `azurerm_arc_machine*`. This solution makes breaking change to current datasource so we have to keep both until 4.0. If we choose this one, I may need to change the datasource first and add this extension with another PR.

**Comment by stephybun** [comment]
@liuwuliuyun - the commercial product name which is user facing and carries more weight than the resource provider's name is Azure Arc (for future reference the words `hybrid` and `connected` are synonymous with `arc` in Azure). For consistency with other Arc specific resources within the provider i.e. `azurerm_arc_kubernetes_cluster` these need to be named/renamed `azurerm_arc_*`.

It's better to fix these now to set the correct precedent.

**Comment by liuwuliuyun** [comment]
Ok, let me start working on that.

**Comment by liuwuliuyun** [comment]
Close this PR due to complete name change on the exsiting resources.

**Pull request closed without merging**